File: content/develop/use-cases/time-series-dashboard/rust/sensor_simulator.rs
//! Simulated sensor definitions and value generation for the dashboard demo.

use rand::Rng;
use std::collections::BTreeMap;

#[derive(Debug, Clone)]
pub struct SensorDefinition {
    pub sensor_id: &'static str,
    pub zone: &'static str,
    pub unit: &'static str,
    pub min_value: f64,
    pub max_value: f64,
    pub baseline: f64,
    pub drift: f64,
    pub spike_chance: f64,
    pub spike_size: f64,
}

impl SensorDefinition {
    pub fn sensor_type(&self) -> &'static str {
        "power_consumption"
    }

    pub fn key(&self) -> String {
        format!("ts:sensor:{}:{}", self.sensor_type(), self.sensor_id)
    }

    pub fn labels(&self) -> BTreeMap<&'static str, &'static str> {
        BTreeMap::from([
            ("site", "demo"),
            ("sensor_type", self.sensor_type()),
            ("sensor_id", self.sensor_id),
            ("zone", self.zone),
            ("unit", self.unit),
        ])
    }
}

#[derive(Debug, Clone)]
pub struct SensorSample {
    pub sensor: SensorDefinition,
    pub value: f64,
}

pub fn sensors() -> Vec<SensorDefinition> {
    vec![
        SensorDefinition {
            sensor_id: "power-1",
            zone: "north",
            unit: "watts",
            min_value: 320.0,
            max_value: 920.0,
            baseline: 480.0,
            drift: 18.0,
            spike_chance: 0.10,
            spike_size: 120.0,
        },
        SensorDefinition {
            sensor_id: "power-2",
            zone: "central",
            unit: "watts",
            min_value: 320.0,
            max_value: 920.0,
            baseline: 560.0,
            drift: 20.0,
            spike_chance: 0.08,
            spike_size: 140.0,
        },
        SensorDefinition {
            sensor_id: "power-3",
            zone: "south",
            unit: "watts",
            min_value: 320.0,
            max_value: 920.0,
            baseline: 640.0,
            drift: 24.0,
            spike_chance: 0.12,
            spike_size: 160.0,
        },
    ]
}

#[derive(Debug, Clone)]
pub struct SensorSimulator {
    sensors: Vec<SensorDefinition>,
    values: BTreeMap<&'static str, f64>,
}

impl SensorSimulator {
    pub fn new(sensors: Vec<SensorDefinition>) -> Self {
        let values = sensors
            .iter()
            .map(|sensor| (sensor.sensor_id, sensor.baseline))
            .collect();

        Self { sensors, values }
    }

    pub fn next_samples(&mut self) -> Vec<SensorSample> {
        let mut rng = rand::thread_rng();
        let mut samples = Vec::with_capacity(self.sensors.len());

        for sensor in &self.sensors {
            let current = *self.values.get(sensor.sensor_id).unwrap_or(&sensor.baseline);
            let drift = random_between(&mut rng, -sensor.drift, sensor.drift);
            let pull = (sensor.baseline - current) * 0.12;
            let mut value = current + drift + pull;

            if sensor.spike_chance > 0.0 && rng.gen::<f64>() < sensor.spike_chance {
                value += random_between(&mut rng, 0.5, 1.0) * sensor.spike_size;
            }

            value = value.clamp(sensor.min_value, sensor.max_value);
            value = (value * 100.0).round() / 100.0;

            self.values.insert(sensor.sensor_id, value);
            samples.push(SensorSample {
                sensor: sensor.clone(),
                value,
            });
        }

        samples
    }
}

fn random_between(rng: &mut impl Rng, min: f64, max: f64) -> f64 {
    min + rng.gen::<f64>() * (max - min)
}

File: content/develop/use-cases/time-series-dashboard/nodejs/sensorSimulator.js
/**
 * Simulated sensor definitions and value generation for the dashboard demo.
 */

class SensorDefinition {
  constructor({
    sensorId,
    zone,
    unit,
    minValue,
    maxValue,
    baseline,
    drift,
    spikeChance = 0,
    spikeSize = 0,
  }) {
    this.sensorId = sensorId;
    this.zone = zone;
    this.unit = unit;
    this.minValue = minValue;
    this.maxValue = maxValue;
    this.baseline = baseline;
    this.drift = drift;
    this.spikeChance = spikeChance;
    this.spikeSize = spikeSize;
  }

  get sensorType() {
    return "power_consumption";
  }

  get key() {
    return `ts:sensor:${this.sensorType}:${this.sensorId}`;
  }

  get labels() {
    return {
      site: "demo",
      sensor_type: this.sensorType,
      sensor_id: this.sensorId,
      zone: this.zone,
      unit: this.unit,
    };
  }
}

const SENSORS = [
  new SensorDefinition({
    sensorId: "power-1",
    zone: "north",
    unit: "watts",
    minValue: 320.0,
    maxValue: 920.0,
    baseline: 480.0,
    drift: 18.0,
    spikeChance: 0.10,
    spikeSize: 120.0,
  }),
  new SensorDefinition({
    sensorId: "power-2",
    zone: "central",
    unit: "watts",
    minValue: 320.0,
    maxValue: 920.0,
    baseline: 560.0,
    drift: 20.0,
    spikeChance: 0.08,
    spikeSize: 140.0,
  }),
  new SensorDefinition({
    sensorId: "power-3",
    zone: "south",
    unit: "watts",
    minValue: 320.0,
    maxValue: 920.0,
    baseline: 640.0,
    drift: 24.0,
    spikeChance: 0.12,
    spikeSize: 160.0,
  }),
];

class SensorSimulator {
  constructor(sensors) {
    this.sensors = sensors;
    this.values = new Map(sensors.map((sensor) => [sensor.sensorId, sensor.baseline]));
  }

  nextSamples() {
    const samples = [];

    for (const sensor of this.sensors) {
      const current = this.values.get(sensor.sensorId);
      const drift = this.randomBetween(-sensor.drift, sensor.drift);
      const pull = (sensor.baseline - current) * 0.12;
      let value = current + drift + pull;

      if (sensor.spikeChance > 0 && Math.random() < sensor.spikeChance) {
        value += this.randomBetween(0.5, 1.0) * sensor.spikeSize;
      }

      value = Math.max(sensor.minValue, Math.min(sensor.maxValue, value));
      value = Math.round(value * 100) / 100;

      this.values.set(sensor.sensorId, value);
      samples.push([sensor, value]);
    }

    return samples;
  }

  randomBetween(min, max) {
    return min + Math.random() * (max - min);
  }
}

module.exports = {
  SensorDefinition,
  SensorSimulator,
  SENSORS,
};
